File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode getKthNode(ListNode node, int k) {
        while(node != null && k > 0){
            node = node.next;
            k --;
        }
        return node;
    }

    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummyHead = new ListNode(0, head);
        ListNode groupPrev = dummyHead;

        while(true) {
            ListNode kthNode = getKthNode(groupPrev, k);
            if(kthNode == null) break;
            ListNode groupNext = kthNode.next;

            // Reverse the nodes untill kthNode
            ListNode curr = groupPrev.next;
            ListNode prev = kthNode.next;

            while(curr != groupNext) {
                ListNode next = curr.next;
                curr.next = prev;
                prev = curr;
                curr = next;
            }

            // (0) -> (1) -> (0) -> (2) -> (1) -> (3) -> (4) 
            ListNode temp = groupPrev.next;
            groupPrev.next = kthNode;
            groupPrev = temp;
        }
        return dummyHead.next;
    }
}

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        
        if not head:
            return head

        curr = head

        while curr:
            curr.next = Node(curr.val, curr.next)
            curr = curr.next.next

        
        # copying random pointers
        curr = head
        while curr:
            if curr.random and curr.next:
                curr.next.random = curr.random.next
            if curr.next:
                curr = curr.next.next

        # Splitting the nodes
        orig_curr = head
        clone_head = head.next

        clone_curr = clone_head

        while orig_curr:
            if orig_curr.next:
                orig_curr.next = orig_curr.next.next

            if clone_curr.next:
                clone_curr.next = clone_curr.next.next

            orig_curr = orig_curr.next
            clone_curr = clone_curr.next

        return clone_head
















